pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java

```diff
@@ -527,4 +605,160 @@ public RebalanceResult rebalance(
       throw new ControllerApplicationException(LOGGER, e.getMessage(), Response.Status.NOT_FOUND);
     }
   }
+
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/tables/{tableName}/state")
+  @ApiOperation(value = "Get current table state", notes = "Get current table state")
+  public String getTableState(
+      @ApiParam(value = "Name of the table to get its state", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "realtime|offline", required = true) @QueryParam("type") String tableTypeStr) {
+    String tableNameWithType = constructTableNameWithType(tableName, tableTypeStr);
+    try {
+      ObjectNode data = JsonUtils.newObjectNode();
+      data.put("state", _pinotHelixResourceManager.isTableEnabled(tableNameWithType) ? "enabled" : "disabled");
+      return data.toString();
+    } catch (TableNotFoundException e) {
+      throw new ControllerApplicationException(LOGGER, "Failed to find table: " + tableNameWithType,
+          Response.Status.NOT_FOUND);
+    }
+  }
+
+  @GET
+  @Path("/tables/{tableName}/stats")
+  @Produces(MediaType.APPLICATION_JSON)
+  @ApiOperation(value = "table stats", notes = "Provides metadata info/stats about the table.")
+  public String getTableStats(
+      @ApiParam(value = "Name of the table", required = true) @PathParam("tableName") String tableName,
+      @ApiParam(value = "realtime|offline") @QueryParam("type") String tableTypeStr) {
+    ObjectNode ret = JsonUtils.newObjectNode();
+    if ((tableTypeStr == null || TableType.OFFLINE.name().equalsIgnoreCase(tableTypeStr)) && _pinotHelixResourceManager
+        .hasOfflineTable(tableName)) {
+      String tableNameWithType = TableNameBuilder.forType(TableType.OFFLINE).tableNameWithType(tableName);
+      TableStats tableStats = _pinotHelixResourceManager.getTableStats(tableNameWithType);
+      ret.set(TableType.OFFLINE.name(), JsonUtils.objectToJsonNode(tableStats));
+    }
+    if ((tableTypeStr == null || TableType.REALTIME.name().equalsIgnoreCase(tableTypeStr)) && _pinotHelixResourceManager
+        .hasRealtimeTable(tableName)) {
+      String tableNameWithType = TableNameBuilder.forType(TableType.REALTIME).tableNameWithType(tableName);
+      TableStats tableStats = _pinotHelixResourceManager.getTableStats(tableNameWithType);
+      ret.set(TableType.REALTIME.name(), JsonUtils.objectToJsonNode(tableStats));
+    }
+    return ret.toString();
+  }
+
+  private String constructTableNameWithType(String tableName, String tableTypeStr) {
+    TableType tableType;
+    try {
+      tableType = TableType.valueOf(tableTypeStr.toUpperCase());
+    } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)